src/analyses/dirty.h

```diff
@@ -17,42 +17,74 @@ Date: March 2013
 #include <unordered_set>
 
 #include <util/std_expr.h>
+#include <util/invariant.h>
 #include <goto-programs/goto_functions.h>
 
 class dirtyt
 {
+private:
+  void die_if_uninitialized() const
+  {
+    INVARIANT(initialized,
+        "Uninitialized dirtyt. This dirtyt was constructed using the default "
+        "constructor and not subsequently initialized using "
+        "dirtyt::build().");
+  }
+
 public:
+  bool initialized;
   typedef std::unordered_set<irep_idt, irep_id_hash> id_sett;
   typedef goto_functionst::goto_functiont goto_functiont;
 
-  explicit dirtyt(const goto_functiont &goto_function)
+  dirtyt():
+    initialized(false)
+  {
+  }
+
+  explicit dirtyt(const goto_functiont &goto_function):
+    initialized(false)
   {
     build(goto_function);
+    initialized=true;
   }
 
-  explicit dirtyt(const goto_functionst &goto_functions)
+  explicit dirtyt(const goto_functionst &goto_functions):
+    initialized(false)
   {
-    forall_goto_functions(it, goto_functions)
-      build(it->second);
+    build(goto_functions);
+    // build(g_funs) responsible for setting initialized to true, since
```

Comment:
This possibly isn't your design - but it would seem to be better for this constructor only to be visible, then you don't need to separately track initialised and have constructors as the only way build this object (of course removing the default constructor). (Possibly mirroring Martin's comment above).

Reply:
Thanks thk123. I believe that this isn't possible:
- `goto_symex_statet` has a `dirtyt` member, which needs to be initialized upon creation of the `goto_symex_statet`.
- However, `goto_symex_statet` is not actually supplied with a goto_functions object until just before it is used for symbolic execution---it does not receive a goto_functions object when it is constructed. The code was previously designed so that `statet`s were only constructed when needed in `goto_symex_main` since they take so much memory, and so they are not a member of `goto_symext`.
- Therefore, we need to initialize the dirtyt only after the statet has been given a goto_functionst.

Does this make sense? I'll add comments clarifying this. Note that the code had previously been doing a similar (but less safe) thing: statet had a _pointer_ to a dirtyt instead of a member, and the dirtyt was a `nullptr` until supplied with a goto_functions. I changed this because we now are actually copying states around and copying objects that contain stale pointers turned out to lead to much more obscure error messages.

Reply:
OK, the only way round this I can think of is having a interface `dirtyt`, an `unitialized_dirtyt` and an `initalized_dirtyt` that inherit from it. Then the unitialized one could just have `INVARIANT(false)`) on all the relevant method calls and the `initlaized_dirtyt` can only be constructed with the right values. 

However, not even convinced myself this is better so I won't block the PR on it.